crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -194,6 +199,46 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
         }
     }
 
+    fn to_isinstance_constraint(&self, classinfo: &Type<'db>) -> Option<Type<'db>> {
+        match classinfo {
+            Type::ClassLiteral(class) => Some(Type::Instance(*class)),
+            Type::Tuple(tuple) => {
+                let mut builder = UnionBuilder::new(self.db);
+                for element in tuple.elements(self.db) {
+                    builder = builder.add(self.to_isinstance_constraint(element)?);
+                }
+                Some(builder.build())
+            }
+            _ => None,
+        }
+    }
+
+    fn add_expr_call(&mut self, expr_call: &ast::ExprCall, expression: Expression<'db>) {
+        let scope = self.scope();
+        let inference = infer_expression_types(self.db, expression);
+
+        let func_name = expr_call.func.as_name_expr();
+
+        if matches!(func_name, Some(name) if name.id() == "isinstance") {
```

Comment:
This is a little naive ;) You need to make sure that it's `builtins.isinstance` from the standard library that's being called, not just any user-defined function that happens to be called `isinstance()`. (It would be great to add a test for this!)

I think what you want to do is add a variant to this enum:

https://github.com/astral-sh/ruff/blob/72c18c8225166f1f08a826c622b9cbcb48331162/crates/red_knot_python_semantic/src/types.rs#L1600-L1606

And then add code here so that we infer `KnownFunction::IsInstance` for `builtins.isinstance()`: https://github.com/astral-sh/ruff/blob/72c18c8225166f1f08a826c622b9cbcb48331162/crates/red_knot_python_semantic/src/types/infer.rs#L778-L783

Reply:
Oops :smile:

Reply:
Another related case that should be handled automatically by Alex's suggested change, and IMO we should add a test for (though perhaps one could argue it's not necessary to support? not sure if mypy/pyright do) is the case where isinstance has been aliased to a different name:

```
>>> myfunc = isinstance
>>> myfunc(1, int)
True
```

Reply:
Mypy supports aliasing `isinstance` by import but not by assignment: https://mypy-play.net/?mypy=latest&python=3.12&gist=14114360a8603492b2f0b992fac6c0c7

Pyright supports both: https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBARgrgSwDYBcEDsDOUEQA5ggo6YaYoCG6AxgKZQXbBhgCwAUBzBSFALwkylGrQ4cAJrWBRgACgAeALihgYAK1rUUASkUcoBnNOZgFAGhzode9obtQQtAG60KSAPooAnnloLt%2BvYGgQYI0twg5lDkILoh9o4ubp4%2BfvLaQA

We should support both, and add tests for both, IMO

Reply:
We now support both (and have tests for both). Thanks!